pkg/status/singletonstatus.go

```diff
@@ -18,65 +18,71 @@ package status
+		if !r || n != 1 {
+			logger.V(4).Info("Workload object should not have singleton status synced, cleaning",
+				"resource", wObjID.Resource, "objectName", wObjID.ObjectName,
+				"requested", r, "nWECs", n)
+			if err := c.reconcileSingletonWObj(ctx, wObjID, false); err != nil {
+				return err
+			}
+		} else {
+			logger.V(4).Info("Workload object should have singleton status synced, updating",
+				"resource", wObjID.Resource, "objectName", wObjID.ObjectName)
+			if err := c.reconcileSingletonWObj(ctx, wObjID, true); err != nil {
+				return err
+			}
 		}
-		return fmt.Errorf("could not get binding %s: %w", bdgName, err)
-	}
-
-	sync := make(map[util.ObjectIdentifier]bool)
-	if err = c.checkSingletonWorkloadInBinding(ctx, *binding, sync, true); err != nil {
-		return err
 	}
+	return nil
+}
 
-	allBdgs, err := c.wdsKsClient.ControlV1alpha1().Bindings().List(ctx, metav1.ListOptions{})
-	if err != nil {
-		return err
-	}
+func (c *Controller) reconcileSingletonByWS(ctx context.Context, ref singletonWorkStatusRef) error {
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Reconciling singleton status due to workstatus changes", "name", string(ref.Name))
+	wObjID := ref.SourceObjectIdentifier
 
-	for _, bdg := range allBdgs.Items {
-		if bdg.Name == binding.Name {
-			continue
-		}
-		if err = c.checkSingletonWorkloadInBinding(ctx, bdg, sync, false); err != nil {
+	requested, nWECs := c.bindingPolicyResolver.GetSingletonReportedStateRequestForObject(wObjID)
+	if !requested || nWECs != 1 {
+		logger.V(4).Info("Workload object should not have singleton status synced, cleaning",
+			"resource", ref.SourceObjectIdentifier.Resource, "objectName", ref.SourceObjectIdentifier.ObjectName,
+			"requested", requested, "nWECs", nWECs)
+		return c.reconcileSingletonWObj(ctx, wObjID, false)
+	} else {
+		logger.V(4).Info("Workload object should have singleton status synced, updating",
+			"resource", ref.SourceObjectIdentifier.Resource, "objectName", ref.SourceObjectIdentifier.ObjectName)
+		wsObj, err := c.workStatusLister.ByNamespace(ref.WECName).Get(ref.Name)
+		if err != nil {
 			return err
 		}
-	}
-
-	for wObjIdentifier, v := range sync {
-		if err := c.reconcileSingletonWObj(ctx, wObjIdentifier, v); err != nil {
+		status, err := util.GetWorkStatusStatus(wsObj)
+		if err != nil {
 			return err
 		}
+		if status == nil {
+			return nil
+		}
+		logger.V(4).Info("Updating singleton status", "workstatus", ref.Name, "wecName", ref.WECName)
+		return updateObjectStatus(ctx, &wObjID, status, c.listers, c.wdsDynClient)
 	}
-	return nil
 }
 
 func (c *Controller) reconcileSingletonWObj(ctx context.Context, wObjID util.ObjectIdentifier, sync bool) error {
 	logger := klog.FromContext(ctx)
-	logger.V(4).Info("Reconciling singleton workload object", "resource", wObjID.Resource, "objectName", wObjID.ObjectName)
+	logger.V(4).Info("Reconciling workload object for singleton status", "resource", wObjID.Resource, "objectName", wObjID.ObjectName)
```

Comment:
If this log message included the value of the `sync` parameter then the other two in this func would not be needed.

Reply:
Included the value of `sync`.

BTW, appreciate for looking at all the logging related details so carefully.